pyface/ui/wx/util/gui_test_assistant.py

```diff
@@ -0,0 +1,272 @@
+#  license included in /LICENSE.txt and may be redistributed only
+#  under the conditions described in the aforementioned license.  The license
+#  is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+#  code based on the traits-enaml.testing.gui_test_assistant
+#
+# ---------------------------------------------------------------------------
+import contextlib
+import threading
+
+import wx
+
+from traits.testing.unittest_tools import UnittestTools
+from traits.testing.unittest_tools import _TraitsChangeCollector as \
+    TraitsChangeCollector
+from pyface.util.guisupport import get_app_wx
+
+from .event_loop_helper import EventLoopHelper, ConditionTimeoutError
+
+
+def find_widget(start, type_, test=None):
+    """Recursively walks the widget tree from widget `start` until it
+       finds a widget of type `type_` (a toolkit widget subclass) that
+        satisfies the provided `test` method.
+
+       Parameters
+       ----------
+       start :
+          The toolkit widget from which to start walking the tree
+       type_ : type
+           A subclass of toolkit widget to use for an initial type filter
+           while walking the tree
+       test : callable
+           A filter function that takes one argument (the current widget being
+           evaluated) and returns either True or False to determine if the
+           widget matches the required criteria.
+
+    """
+    if test is None:
+        test = lambda widget: True
+    if isinstance(start, type_):
+        if test(start):
+            return start
+    for child in start.GetChildren():
+        widget = find_widget(child, type_, test=test)
+        if widget:
+            return widget
+    return None
+
+
+def print_widget_tree(widget, level=0):
+    """ Debugging helper to print out the widget tree starting at a
+    particular `widget`.
+
+    Parameters
+    ----------
+    widget :
+        The root widget in the tree to print.
+    level : int
+        The current level in the tree. Used internally for displaying the
+        tree level.
+    """
+    level = level + 4
+    if level == 0:
+        print
+    print ' '*level, widget
+    for child in widget.GetChildren():
+        print_widget_tree(child, level=level)
+    if level == 0:
+        print
+
+
+class GuiTestAssistant(UnittestTools):
+
+    def setUp(self):
+        # Make sure that we have a running application.
+        self.gui_app = get_app_wx()
+        wx.Log.SetActiveTarget(wx.LogStderr())
+        self.event_loop_helper = EventLoopHelper(gui_app=self.gui_app)
+
+    def tearDown(self):
+        windows = wx.GetTopLevelWindows()
+        titles = []
+        if len(windows) > 0:
+            for window in windows:
+                titles.append(window.GetTitle())
+
+        def _cleanup():
+            for window in wx.GetTopLevelWindows():
+                window.Destroy()
+                wx.WakeUpIdle()
+
+        wx.CallLater(50, _cleanup)
+        self.gui_app.MainLoop()

```

Comment:
Could we do `self.event_loop_helper.event_loop_until_condition(<callable-that-returns-true-when-all-windows-are-closed>)` instead?  There is a danger of this MainLoop call hanging there forever.

Reply:
The wx event loop exists only when all the windows are closed. Event_loop_until_codition does the same so there is no advantage using it.

Reply:
oops I was wrong

Reply:
So the right thing here is to use a wx.Timer 

something like
```
handler = wx.EvtHandler()
kill_timer = wx.Timer(handler) 
handler.Bind(wx.EVT_TIMER, lambda event:self.gui_app.Exit(), kill_timer) 
kill_timer.Start(30000), True) 
try:
self.gui_app.MainLoop()
finally:
      kill_timer.Stop()
```

Reply:
Does the timer have the same effect as `wx.CallLater(30000, self.gui_app.Exit)`?

Reply:
It is not the same. Since with the time we can deactivate it while the CallLater will be called later probably while on another is running.

Reply:
Ah!  Good point!